### ISSUE TITLE
Exception thrown when listening to caret bounds property and the text area is resized

### ISSUE THREAD

**Issue by garybentley** [FXMisc/RichTextFX#842]
I added a listener to the caret bounds property and get the following exception (truncated) when I resize the text area.

```
java.lang.IndexOutOfBoundsException: Index 14 out of bounds for length 14
        at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:64)
        at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:70)
        at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:248)
        at java.base/java.util.Objects.checkIndex(Objects.java:372)
        at java.base/java.util.ArrayList.get(ArrayList.java:458)
        at javafx.base/com.sun.javafx.collections.ObservableListWrapper.get(ObservableListWrapper.java:89)
        at javafx.base/com.sun.javafx.collections.VetoableListDecorator.get(VetoableListDecorator.java:306)
        at javafx.graphics/javafx.scene.Parent.layout(Parent.java:1210)
        at javafx.graphics/javafx.scene.Parent.layout(Parent.java:1213)
        at org.fxmisc.flowless.VirtualFlow.getCellIfVisible(VirtualFlow.java:220)
        at org.fxmisc.richtext.GenericStyledArea.getCaretBoundsOnScreen(GenericStyledArea.java:1033)
        at org.fxmisc.richtext.CaretNode.lambda$new$6(CaretNode.java:232)
        at org.reactfx.value.Val$3.computeValue(Val.java:722)
        at org.reactfx.value.ValBase.getValue(ValBase.java:17)
        at org.reactfx.value.SuspendableValWrapper.getValue(SuspendableValWrapper.java:23)
        at org.reactfx.value.ChangeListenerWrapper.accept(Val.java:784)
        at org.reactfx.util.AbstractReducingStreamNotifications.lambda$head$0(NotificationAccumulator.java:248)
        at org.reactfx.ObservableBase.notifyObservers(ObservableBase.java:68)
        at org.reactfx.ObservableBase.notifyObservers(ObservableBase.java:57)
        at org.reactfx.SuspendableBase.handleEvent(SuspendableBase.java:82)
        at org.reactfx.util.NonAccumulativeStreamNotifications.lambda$head$0(NotificationAccumulator.java:134)
        at org.reactfx.ObservableBase.notifyObservers(ObservableBase.java:68)
        at org.reactfx.ObservableBase.notifyObservers(ObservableBase.java:57)
        at org.reactfx.ProperEventStream.emit(ProperEventStream.java:18)
```

Any idea what the issue is?

**Comment by garybentley** [comment]
To give more context here.  I also have a Paragraphic Graphic Factory set on the area.  The factory is showing nodes in the graphic that are related to positions in the text.

If I don't have a caret bounds listener then the graphic factory works ok.  If I do have a graphic factory added then when the caret bounds property is being updated the area tries to do a layout, which triggers the graphic factory to layout which tries to get the character positions which then causes the exception because the text doesn't seem to have been laid out fully.

You can see this behavior with the exception below:

```
java.lang.NullPointerException
	at org.fxmisc.richtext.GenericStyledArea.getParagraphBoundsOnScreen(GenericStyledArea.java:1528)
	at java.base/java.util.Optional.map(Optional.java:265)
	at org.fxmisc.richtext.GenericStyledArea.getParagraphBoundsOnScreen(GenericStyledArea.java:1028)
	at com.quollwriter.ui.fx.components.TextEditor.getBoundsForPosition(TextEditor.java:381)
	at com.quollwriter.ui.fx.panels.ParagraphIconMargin.getItemsForPosition(ParagraphIconMargin.java:299)
	at com.quollwriter.ui.fx.panels.ParagraphIconMargin.layoutStructureItems(ParagraphIconMargin.java:415)
	at com.quollwriter.ui.fx.panels.ParagraphIconMargin.layoutChildren(ParagraphIconMargin.java:570)
	at javafx.graphics/javafx.scene.Parent.layout(Parent.java:1206)
	at javafx.graphics/javafx.scene.Parent.layout(Parent.java:1213)
	at javafx.graphics/javafx.scene.Parent.layout(Parent.java:1213)
	at javafx.graphics/javafx.scene.Parent.layout(Parent.java:1213)
	at org.fxmisc.flowless.VirtualFlow.layoutChildren(VirtualFlow.java:257)
	at javafx.graphics/javafx.scene.Parent.layout(Parent.java:1206)
	at org.fxmisc.flowless.VirtualFlow.getCellIfVisible(VirtualFlow.java:220)
	at org.fxmisc.richtext.GenericStyledArea.getCaretBoundsOnScreen(GenericStyledArea.java:1033)
	at org.fxmisc.richtext.CaretNode.lambda$new$6(CaretNode.java:232)
	at org.reactfx.value.Val$3.computeValue(Val.java:722)
	at org.reactfx.value.ValBase.getValue(ValBase.java:17)
	at org.reactfx.value.ValBase.newObserver(ValBase.java:40)
	at org.reactfx.value.ValBase.newObserver(ValBase.java:8)
	at org.reactfx.ObservableBase.addObserver(ObservableBase.java:110)
	at org.reactfx.ObservableBase.observe(ObservableBase.java:100)
	at org.reactfx.value.Val.observeInvalidations(Val.java:61)
	at org.reactfx.value.Val$1.observeInputs(Val.java:104)
	at org.reactfx.ObservableBase.addObserver(ObservableBase.java:108)
	at org.reactfx.ObservableBase.observe(ObservableBase.java:100)
	at org.reactfx.EventStream.subscribe(EventStream.java:53)
	at org.reactfx.SuspendableBase.observeInputs(SuspendableBase.java:49)
	at org.reactfx.ObservableBase.addObserver(ObservableBase.java:108)
	at org.reactfx.value.Val.addInvalidationObserver(Val.java:52)
	at org.reactfx.value.Val.addListener(Val.java:70)
	at com.quollwriter.ui.fx.panels.ChapterEditorPanelContent.lambda$init$7(ChapterEditorPanelContent.java:218)

```

My graphic factory is the "ParagraphIconMargin" in the exception above.

**Comment by Jugen** [comment]
I've had a look at this and I need more info or preferably some code please that I can run to replicate the problem.

**Comment by garybentley** [comment]
I've been trying all day to create a reproducible example.  But I've failed.  My problem is that the exception doesn't hit my code so I've had to just trial and error things, commenting out code and seeing if the problem goes away.  I think I've now got a workaround but the exception still happens once when the area becomes very narrow, but I can also workaround that by preventing the area from becoming too narrow.

I think the problem was related to me adding too many nodes to the paragraph graphic at the same location, but I'll be damned if I can determine why.
